Add dynarec article (#25)
emudev-org/emudev.org: _posts/2021-02-01-Dynarec.md

recompilers, but fewer have implemented one. A good way to start is through one
of the beginner tutorials available on the Getting Started page. But while the
simple dynarecs described by these tutorials may be satisfactory for some
systems, a lot more performance remains on the table, and in between beginner
content and the cutting edge there are many widely known optimizations that can
be taken for granted by more experienced developers. This article will cover
some of these baseline optimizations, with a focus on getting the most
performance for the least code complexity.

## Introduction

It is important to note before we start that there are some inherent limitations
to what this guide can help you with. Obviously, it will not cover bottlenecks
outside of the JIT, of which there are plenty in most full system emulators. As
you optimize the bottlenecks will naturally shift - remember to profile changes
and target the parts that actually need improvement, which is frequently
scheduling and synchronization, peripherals, or graphics rather than actual
instruction execution - so it probably doesn't make sense to implement all of

> these JIT changes at once, without investigating other components in between.

Reviewer: You kind of already imply this, but it might be wise to explicit state something along the lines of:

"As with all optimisation tasks, it's important to actually profile your emulator and determine the JIT is currently the bottleneck before proceeding with optimisations"

as a programmer, I know I always forget this and could do with extra explicit reminders from time to time.

which will provide your particular use case the most bang for the buck. Knowing
more about the actual characteristics of the code you are executing - what
addresses it jumps to, which instructions it executes most often, etc. will pay
off far more than brute force application of a list of techniques. For context,
my own experience was mostly gained through an N64 emulator, which has two MIPS
III guests on x86_64 desktop hosts. I've also discussed JIT design with inolen
and Merrymage, creators of redream and dynarmic respectively, who both have far
more experience in this subject than I.

We start with the following pseudocode, representing the control flow of an
unoptimized JIT translating one basic block at a time. We won't go into the
details of compilation, because for the simplest binary translators (ones that
don't call LLVM) the overwhelming majority of overhead comes not from
inefficient code or slow compilation, but from the surrounding basic block loop.

```
// outer loop in C++
// lookup = unordered_map<uint32_t, Block>
while (time_slice > 0) {
  pc = check_interrupts(pc);
  block = lookup[pc];
  if (block) pc = block();

>   else lookup[pc] = compile(pc);

Reviewer 2: i  guess this assumes that `compile` also runs runs compiled code? (like `block()`) ?
if so, to be more clear this could be
```
if (!block) lookup[pc] = compile(pc);
pc = block();
```

}

// in every emitted block
save_host_regs(), load_guest_regs()
// translate instructions, set next_pc
save_guest_regs(), load_host_regs()
return next_pc
```

## Indirect Branch Lookup

Our first and most important optimization is to move the most common path
through this outer loop into assembly. Most of the time, we won't need to handle
an interrupt or scheduler event, so blocks only need to verify that they do not

> need to exit the JIT, then lookup the address of the translation for next block

Reviewer 2: ```suggestion
need to exit the JIT, then lookup the address of the translation for the next block
```

and jump to it - if the exit checks do not pass or the next block has not yet
been compiled, only then do we have to incur the overhead of saving and
restoring registers. Doing a hash table lookup in assembly can be expensive, not

> to mention difficult to implement, so one way to simplify is to replace it with

Reviewer 3: You don't need to access the full-blown hash-table here to guarantee you get a yes/no answer. You can access an intermediate, easy-to-implement hash-table, and if that lookup fails you can do a full lookup.

Reviewer 3: Possibly a bit too much to mention here but I feel I should note that most "unpredictable" jumps actually only jump to a small set of addresses, which can potentially be sped up with a local tiny hash-table of a handful of elements.

Reviewer 3: An optimization that is worth mentioning here is the [return stack buffer](https://en.wikipedia.org/wiki/Branch_predictor#Prediction_of_function_returns) (i.e. return address prediction).

Author: I didn't write about these largely because I haven't implemented them myself (so I don't know the details well) but I should at least mention them. Do you know of any good articles I can link to about the local hash tables and RSBs? I'm especially curious about how invalidation works for both.

Reviewer 3: [Fiora's dolphin article in PoC||GTFO 0x06](https://www.alchemistowl.org/pocorgtfo/pocorgtfo06.pdf#page=11) mentions RSBs. Though dolphin's implementation also ends up abusing the host RSB which I feel is unnecessary.

I have [documentation of my implementation](https://github.com/MerryMage/dynarmic/blob/master/docs/ReturnStackBufferOptimization.md) though it's out of date (I no longer scan the entire RSB).

Invalidation for an RSB is done by clearing it completely. Invalidating an RSB at any point does not result in incorrect behavior, only loss of performance gain.

Reviewer: I find it interesting that you get good performance scanning a linear list. Do you have metrics for how often the later entries are hit? 

I would have thought that if the first entry doesn't hit, the probability of the next 7 hitting would be too low to be bothered worrying about. 

> Though dolphin's implementation also ends up abusing the host RSB which I feel is unnecessary.

I like Dolphin's solution, it's simple as you only push one extra value to the host stack (the expected guest return PC). then to emulate a `ret` you check if the value on the stack is equal to the expected guest PC and execute a host `ret`.

I would think (though I haven't exactly benchmarked) that the performance advantages of triggering the host CPU's return prediction would be worth the extra complexity of the abort case.

Dolphin's abort case simply resets the host stack to zero, declaring bankruptcy on all future predictions (and preventing the host stack growing forever in edge cases)

Reviewer: > You don't need to access the full-blown hash-table here to guarantee you get a yes/no answer. You can access an intermediate, easy-to-implement hash-table, and if that lookup fails you can do a full lookup.

For example, FEX has a quick-and-dirty "hash-function" for our L1 cache that simply truncates the PC to 12 bits, knocking off 4 bits at the bottom and the rest from the top. This cuts the size of the "hash-table" down to 16MB (8 bytes expected PC, 8 bytes block address) and can be implemented in just a few assembly instructions, which we append to any block that does an indirect call.   
If we miss we fall back to a proper lookup in c++ code. 

The 16MB size is just a placeholder. Eventually we will do some profiling to find the optimal trade-offs between hash-table size, hash function complexity and hit rate for the best performance. I'd be surprised if the optional size for the L1 hash-table is more than 16KB.

Author: I've added some notes on what you've described below. I think these indirect branch optimizations would be a great topic for a future long form article by someone who has implemented them. I'm sure a lot of others would appreciate your writing.

Reviewer 3: @phire

> I would have thought that if the first entry doesn't hit, the probability of the next 7 hitting would be too low to be bothered worrying about.

I used to get a decent hit rate back on multitheaded workloads with multiple threads while sharing a per-core RSB, but since switching to per-emulated-thread RSB things became more sane.

a flat lookup table of pointers, one for each possible guest target address.
With this we may be able to replace a function return, hash table lookup, and
function call to the next block, each of which may take a large number of
instructions, with a load of the lookup table base address, a read of the
appropriate element of the lookup table using the guest address, and an indirect
branch to the pointer contained in that element.

Of course, this approach wastes a lot of memory compared to a hash table, but
there are several ways to reduce it. In many consoles, instructions must be
4-byte aligned. Of course, if the guest has virtual memory, you probably do the

> physical translation first. If code must be executed from ROM, only ROM

Reviewer: Only for system level emulators. Userspace emulators get to skip this uglyness.

pre-filling the lookup table with the address of a common code fragment that
calls the recompiler, so the single indirect branch can be used for both the
common case of jumping directly to the next block, and the uncommon case of
compiling it. Another source of things that can be moved to "far code" is
exceptions, including interrupts and page faults. These often require storing
the current pc, reading or modifying status registers, and jumping to an
exception handler, though this code is skipped over in most blocks. In general,
it makes sense to find ways of moving as much infrequently used computation as
possible to far code, by for example by forcing the time slice to 0 when an
interrupt occurs, or lazily calculating counter registers from scheduler time,
instead of incrementing it at the end of blocks. This general pattern of
reducing overhead on the common path, at the expense of more logic elsewhere, is
the essence of most JIT optimizations.

## Static Branch Linking

So far, these optimizations have been built on assumptions of the emulator path,
but information available for specific blocks at compile time can be used for

> further performance improvements. The most obvious and important if these is

Reviewer 3: ```suggestion
further performance improvements. The most obvious and important of these are
```

system would take a lot of space and use a lot of overhead for most systems, but
it can make sense when there is very limited memory, or specific pages are known
to frequently contain both data and code. In the latter case access to those
pages would go through a special handler, and the bitmaps would only cover that
subset of memory.

For code that truly is frequently overwritten though, even fine-grained bitmaps
won't help. In that case the simplest option is to emit code at the beginning of
every block that computes a checksum of the addresses it would fetch, and verify
that this is the same as it was at compile time. If they are different, we must
immediately exit the JIT, invalidate the block, and compile a new one. On x86,
the hardware crc32 instructions may be useful for the checksum portion of this
task. In many cases frequently invalidated code is also reused later, so it may
be wise to compute the checkum up to the next branch, instead of to where the
end of the block was at compile time. This way the checksum can be used to index
into a backup content-based block lookup table which can replace the newly

> invalidated block without actually recompiling.

Reviewer 3: When this happens, in order to reduce the code that needs to be checksummed, you can split a basic block so that the frequently overwritten code is its own BB.

Self-modifying code tends to come in two forms:
- Replacement of large number of bytes / pages at runtime
- Overwriting a single instruction very frequently

The latter benefits from the above approach.

Author: Do you have more details somewhere about how you distinguish the two cases you mention when detecting writes, and how you'd lookup old blocks for reuse after you split the block?

Reviewer 3: I think I should write something specifically about invalidation in the future, since it's interesting to optimize.

Reviewer: Regarding replacement of large numbers of pages, it can be useful very to hook the the system APIs which are commonly used to load new code.

For example, most "real" operating systems use mmap to map new executable pages into memory. You can hook munmap of executable pages to directly invalidate all code being unloaded (before a new mmap call might map new code at the same place).

For consoles, it might be worth having large DMAs from disk/cart check if they are overlapping executable pages before writing. Or detecting overwrite of the first protected page via a segfault and then falling back to a full overlap check.

instruction, on many systems we can take simplifying the fast path even farther
using a combination of host page fault write detection and backpatching, usually
known as "fastmem". This reduces the initial emitted code for memory accesses to
a direct read or write to an address-space-sized array, where pages that require
more complex code to access - such as TLB-mappable regions and MMIO - are
protected by the host MMU. Accesses to those regions then generates a page
fault, the handler for which can use both the particular protected address
accessed, and the address of the faulting instruction, to replace the direct
write with more complex or specialized code as needed - for example, a call to
an assembly stub implementing inline page table lookup if the write was of a
TLB-mappable region, a call to a stub checking the address against a bitmap if
the write was to a page containing code, or a checked direct call to an MMIO
handler if the write was to an MMIO register. This kind of patching not only
improves RAM access speed, but also makes MMIO polling faster, by making use of
the fact that specific instruction addresses usually access only one or the

> other.

Reviewer: Might read better as "usually access only RAM or only MMIO"

so that the write can be undone when the target is invalidated. It can then jump
to the host address of the next block to continue execution as usual. If the
next block has not been compiled yet, then we may have to skip the writing and
exit the JIT to compile, then wait for the next time the block is called to
perform any linking.

If the linked block ends in a guest conditional branch, the linker function may
be called a second time from the same site, when the other arm of the branch is
taken. Since there are only two arms, we can assume the linker function cannot
be called a third time from the same site, so we can now write an unconditional
branch to the next host address over that linker call site, assuming there is
enough memory space there before the end of the block. The easiest way to check
this condition is to check whether the host address passed to the linker points
to nops, or if these have already been overwritten. The following pseudocode
shows what the epilogue of a block ending in a guest conditional branch might
look like at the 3 stages in its "life cycle":

```
// initial state of epilogue
START:

> nops;

Reviewer 4: Do you think it's worth mentioning checking the time slice/downcount here? I suppose it depends on what you're emulating, but otherwise an infinite loop in the guest code will also lead to an infinite loop on the host.

reducing overhead on the common path, at the expense of more logic elsewhere, is
the essence of most JIT optimizations.

The remaining methods in this section I don't have direct experience
implementing, but are worth mentioning for those emulating systems with other
characteristics. If your address space is very large, as might be the case for
modern systems with GBs of RAM, it may not be possible or performant to use the
large lookup table method. Instead we can find ways to perform the expensive
part of hash table lookup as infrequently as possible. In the most general case,
we can make use of the fact that in most reasonably loaded hash tables, the
first slot the key maps to contains the value we want - that is, most accesses
do not require probing. Therefore, we can put only the code for hashing and the
initial lookup in assembly, and exit the JIT only if probing is required. This
is not only easier to write in assembly, but also keeps infrequently used code
from using up cache space. It's also possible to augment the global hash table

> with a smaller, possibly block-local, hash or address table for unpredictable

Reviewer: > possibly block-local, hash or address table

Hmmmm.......... I hadn't actually thought of doing a per-block hash-table. 

There is logic to that, each table is likely small. You could even make a dynamically sized hash function that gets back-patched and resized on misses. 

Start off with a single-entry (because a lot of indirect branches will always hit the same address) and work your way up to larger sizes. 

One downside is that you duplicate cache entries when a lot of indirect branch locations are jumping to the same set of indirect branches (a computed goto interpreter would be the worst case here, with say 256 jump-sites all indirect-jumping to the same 256 jump-sites).

There are various workarounds to this. If a local hash-table exceeds a certain size, it could be updated to point at a global hash-table. Another option is explictly detecting jumptables.
